design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+use the new APIs without any other system. There are three alternatives: 
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. If the columns contain `ROWKEY`, the
+value will be verified against the key schema but will not populate the corresponding field in the 
+value. Populating columns with `ROWTIME` will not be supported.
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Key Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT INTO`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. To all for all levels of flexibility, `INSERT INTO` should allow
+for all cases that kafka supports. Therefore, `ROWKEY` will always be available in the list of
```

Comment:
> To all for all levels of flexibility, `INSERT INTO` should allow for all cases that kafka supports

I disagree. We should model the current state that KSQL has of the world, which is that in the case of 

```
CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
```

the key of the message _is_ `id`. That sometimes people have messages in which this is not the case is a cause of woe and gnashing of teeth, but we should not build `INSERT INTO` such that it can support such state of affairs. We're building `INSERT INTO` to help with development and sandboxing with KSQL and we don't want to give people the gun with which to shoot themselves in the foot. 

If I run 

```
CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
```

I expect to then run 

INSERT INTO bar (id, foo) VALUES ('BAR',2);

and have a Kafka message that looks like this: 

```
{"key": "BAR", "value": {"id": "BAR", "foo": 2}}
```

Reply:
> To all for all levels of flexibility, INSERT INTO should allow for all cases that kafka supports

Wow I can't believe I typed that and didn't catch that in my proofing...

The original intention here was to provide a way to produce a record with a key and a null value (i.e. a tombstone) even if the key appeared in the value, but after some other discussions I decided instead to introduce the `DELETE` statement so I think we can indeed simplify it  by auto-populating the `ROWKEY`.

Reply:
See most updated section.